### PR TITLE
Catch exception when trying to modify already sent headers

### PR DESCRIPTION
**Pull request by jborut**

### CODE REVIEW

**Comment by bripkens** [comment]
How can this happen? Do you have a stacktrace?

**Comment by jborut** [comment]
https://instana.zendesk.com/agent/tickets/4313

{"@timestamp":"2018-09-10T08:58:56.519Z","app":"TheProjectName","level":"error","message":"A promise was rejected but the error wasn't handled:","errMsg":"Can't set headers after they are sent.","errStack":"\"NetworkingError: Can't set headers after they are sent.\\n at validateHeader (_http_outgoing.js:491:11)\\n at ClientRequest.setHeader (_http_outgoing.js:498:3)\\n at /the-project-name/node_modules/instana-nodejs-sensor/src/tracing/instrumentation/httpClient.js:83:21\\n at Namespace.run (/the-project-name/node_modules/instana-nodejs-sensor/src/tracing/clsHooked/context.js:108:5)\\n at Object.request (/the-project-name/node_modules/instana-nodejs-sensor/src/tracing/instrumentation/httpClient.js:43:12)\\n at features.constructor.handleRequest (/the-project-name/node_modules/aws-sdk/lib/http/node.js:42:23)\\n at executeSend (/the-project-name/node_modules/aws-sdk/lib/event_listeners.js:329:29)\\n at Request.SEND (/the-project-name/node_modules/aws-sdk/lib/event_listeners.js:343:9)\\n at Request.callListeners (/the-project-name/node_modules/aws-sdk/lib/sequential_executor.js:105:18)\\n at Request.emit (/the-project-name/node_modules/aws-sdk/lib/sequential_executor.js:81:10)\\n at Request.emit (/the-project-name/node_modules/aws-sdk/lib/request.js:683:14)\\n at Request.transition (/the-project-name/node_modules/aws-sdk/lib/request.js:22:10)\\n at AcceptorStateMachine.runTo (/the-project-name/node_modules/aws-sdk/lib/state_machine.js:14:12)\\n at /the-project-name/node_modules/aws-sdk/lib/state_machine.js:26:10\\n at Request.<anonymous> (/the-project-name/node_modules/aws-sdk/lib/request.js:38:9)\\n at Request.<anonymous> (/the-project-name/node_modules/aws-sdk/lib/request.js:685:12)\""}

**Comment by basti1302** [comment]
Thanks @jborut 

@bripkens: It's a file upload and something with HTTP 100 continue, so that might be something that we never tested.

I'm not 100% sure this approach is the best solution, though. I think I would like to reproduce this first to get a better understanding.

**Comment by basti1302** [comment]
I've extended the PR "a bit" ;-)

An explanation of the root cause is here: https://github.com/instana/nodejs-sensor/pull/81/files#diff-66dc0be4c316b958c2231e912c234e47R209

The fix for the issue the customer is seeing is in this commit: https://github.com/instana/nodejs-sensor/pull/81/commits/332697018bc86c1038f3e3d7f7719e5ab15057d7, although unfortunately this commit also contains another fix for a smaller issue I stumbled over when investigating this - for https calls, we would run our instrumentation code twice.

I was able to reproduce the original issue with an upload to S3 via `aws-sdk`. From there, I was able to create a much simpler reproduction case involving the "Expect: 100-continue" header. I left the original reproduction case in place as a skipped test, just for shits and giggles.

The third commit in this PR is somewhat unrelated, it makes sure `http.get` is also correctly traced (in addition to `http.request`).

@jborut: You are welcome to re-review this change, but since you are not too familiar with the Node.js sensor, a review by @bripkens might also be a good idea.

**Comment by basti1302** [comment]
Well, this really escalated a bit. Now this PR also adds support for passing the URL in as a WHATWG URL object, which is supported beginning of Node.js 7.

**Comment by basti1302** [comment]
Travis complains, gotta look into that.

File: src/tracing/instrumentation/httpClient.js [hunk 1]
```diff
@@ -66,23 +111,30 @@ function instrument(coreModule) {
...
+      }
+
       try {
-        clientRequest = originalRequest.call(coreModule, opts, responseListener);
+        var instanaHeadersHaveBeenAdded = tryToAddHeadersToOpts(options, span);
+        clientRequest = originalRequest.apply(coreModule, originalArgs);
       } catch (e) {
         // synchronous exceptions normally indicate failures that are not covered by the
         // listeners. Cleanup immediately.
         throw e;
```
[the remark this answers is not in the document]
Comment (bripkens) [reply src/tracing/instrumentation/httpClient.js]: This looks as if we had some cleanup logic here which no longer exists. Was this removed by accident?

File: test/tracing/httpClient/upload/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# What's this file doing here?
+
+It's used to test file uploads with Instana's HTTP client instrumentation. It is an excerpt of a complete recording of Verdi's "Messa da Requiem" by the Orchestre national de France, conducted by Igor Markevitch, recorded on May 7th, 1959 in the Théâtre des Champs-Elysées, Paris and published in 1959. The excerpt is section 7.2, "Libera me/Dies irae". The complete recording has been obtained from [IMSLP](https://imslp.org/wiki/Requiem_(Verdi,_Giuseppe)). The requiem itself as well as this particular recording are in the public domain.
```
Comment (bripkens) [inline test/tracing/httpClient/upload/README.md]: 👍

File: src/tracing/instrumentation/httpClient.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (basti1302) [reply src/tracing/instrumentation/httpClient.js]: Indeed it was, with the switch from async-wrap to cls-hooked here: 5dbf7ab7db13ed631e7ae9f5c0c539a541b13736


```
-    try {
-      clientRequest = originalRequest.call(coreHttpModule, opts, responseListener);
-    } catch (e) {
-      // synchronous exceptions normally indicate failures that are not covered by the
-      // listeners. Cleanup immediately.
-      hook.postAndDestroySimulated(uid);
-      throw e;
+      return clientRequest;
     }
```

Okay then, I think this try-catch can be simply removed completely.

[the remark this answers is not in the document]
Comment (basti1302) [reply src/tracing/instrumentation/httpClient.js]: @bripkens agree?

[the remark this answers is not in the document]
Comment (bripkens) [reply src/tracing/instrumentation/httpClient.js]: Actually I think we should execute the span cleanup in this location.

**Pull request merged**